src/main/java/edu/ucla/library/avpairtree/verticles/WaveformVerticle.java

```diff
@@ -0,0 +1,223 @@
+            if (0 == exitValue) {
+                LOGGER.debug(cmdResult);
+            } else {
+                LOGGER.error(cmdResult);
+                aPromise.fail(cmdResult);
+
+                return;
+            }
+        } catch (final IOException | InterruptedException details) {
+            final String startErrorMsg = LOGGER.getMessage(MessageCodes.AVPT_016, cmdline, details);
+
+            LOGGER.error(startErrorMsg);
+            aPromise.fail(details);
+
+            return;
+        }
+
+        mySourceDir = config.getString(Config.SOURCE_DIR);
+
+        // Make sure that configuration and credentials for AWS S3 have been provided
+
+        if (myAwsDefaultRegion == null) {
+            configErrorMsg = LOGGER.getMessage(MessageCodes.AVPT_018);
+
+            LOGGER.error(configErrorMsg);
+            aPromise.fail(configErrorMsg);
+
+            return;
+        }
+
+        if (System.getenv(ProfileProperty.AWS_ACCESS_KEY_ID.toUpperCase()) == null ||
+                System.getenv(ProfileProperty.AWS_SECRET_ACCESS_KEY.toUpperCase()) == null) {
+            configErrorMsg = LOGGER.getMessage(MessageCodes.AVPT_017);
+
+            LOGGER.error(configErrorMsg);
+            aPromise.fail(configErrorMsg);
+
+            return;
+        }
+
+        myS3Bucket = config.getString(Config.AUDIOWAVEFORM_S3_BUCKET);
+
+        if (myS3Bucket == null) {
+            configErrorMsg = LOGGER.getMessage(MessageCodes.AVPT_020);
+
+            LOGGER.error(configErrorMsg);
+            aPromise.fail(configErrorMsg);
+
+            return;
+        }
+
+        myS3Client = S3AsyncClient.builder().region(Region.of(myAwsDefaultRegion)).build();
+
+        vertx.eventBus().<CsvItem>consumer(getClass().getName()).handler(this::handle);
+
+        aPromise.complete();
+    }
+
+    /**
+     * Transforms the source audio file at the given path into audiowaveform data, uploads that data to S3, and replies
+     * to the message with the URL for the data. If either the transformation or upload fails, sends back error details.
+     *
+     * @param aMessage A message with the file path of the audio file to transform
+     */
+    private void handle(final Message<CsvItem> aMessage) {
+        try {
+            final CsvItem csvItem = aMessage.body();
+            final Path audioFilePath = AvPtUtils.getInputFilePath(csvItem, mySourceDir);
+
+            audiowaveform(audioFilePath).onSuccess(s3ObjectData -> {
+                final String ark = csvItem.getItemARK();
+                final String s3ObjectKey = StringUtils.format(S3_OBJECT_KEY_TEMPLATE, ark);
+                final PutObjectRequest req = PutObjectRequest.builder().bucket(myS3Bucket).key(s3ObjectKey).build();
+                final AsyncRequestBody body = AsyncRequestBody.fromByteBuffer(s3ObjectData);
+
+                // Store the audiowaveform data on S3
+                myS3Client.putObject(req, body).whenComplete((resp, err) -> {
+                    if (resp != null) {
```

Comment:
Is a non-null response really the same thing as success? There are no other response codes other than 2xx?

Reply:
Not asking for a change here, because I'd want to pull out the AWS lib once the other lib is updated, but just curious.

Reply:
I don't have a definitive answer on this yet, but [I used the same pattern used by the SDK's docs on asynchronous programming](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/asynchronous.html), assuming that they are showing the best practice.

Reply:
Interesting. Thanks for the link. I definitely like Vert.x's async API better.